### ISSUE TITLE
Unrecognized valid DKIM keys

### ISSUE THREAD

**Issue by pik-256** [hmailserver/hmailserver#382]
Hmailserver cannot recognize DKIM keys with spaces in base64 encoded string. It logs a debug notice:

`DKIM: Unable to base64 decode public key found in DNS record. Key: MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxOsFyzzmGNCLr2ay1xBqHVEFHOPS3GFL9Ys7KMgd8IC58SAXE75Gq71oEkAxrDDuOvCk60PokCarmi3RQf3/+VYGxS90GGF4+3T3mfCPmq3qBZr7vqvhxu199O7H41VbV3Kqmp20j5RF0zoRWCyxUDsKOvjg1YKy283hmvi5h0DHb7hZbdCn7sllzl4FtPn8+nZx uI4jlEECIL8KFY6CnBUNxoBhdDmkGXPwhIAakHW7/KtBmbPXuIORVaNprH4yOjNcq2xxOOqnL1AWYjNheZpo0TwhBS4IX0qWlbk2OlZbsudHJZlVqbpp3Wwkoly/Fu4Jha+/zVU9+HULLJdYOwIDAQAB` 
to a message with a signature header:

`DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed; d=mail.wyborcza.pl; s=m3; .....`

The DNS answer section for m3._domainkey.mail.wyborcza.pl looks like this

`
m3._domainkey.mail.wyborcza.pl, CNAME, m3.dkim.sare.pl
m3.dkim.sare.pl, TXT, "v=DKIM1;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxOsFyzzmGNCLr2ay1xBqHVEFHOPS3GFL9Ys7KMgd8IC58SAXE75Gq71oEkAxrDDuOvCk60PokCarmi3RQf3/+VYGxS90GGF4+3T3mfCPmq3qBZr7vqvhxu199O7H41VbV3Kqmp20j5RF0zoRWCyxUDsKOvjg1YKy283hmvi5h0DHb7hZbdCn7sllzl4FtPn8+nZx uI4jlEECIL8KFY6CnBUNxoBhdDmkGXPwhIAakHW7/KtBmbPXuIORVaNprH4yOjNcq2xxOOqnL1AWYjNheZpo0TwhBS4IX0qWlbk2OlZbsudHJZlVqbpp3Wwkoly/Fu4Jha+/zVU9+HULLJdYOwIDAQAB;
`

This is multi text response with spaces in a string (on the beginning of the second string). Hmailserver treats such keys as invalid. However, this is a valid key as it states in [RFC4871](https://datatracker.ietf.org/doc/html/rfc4871#section-3.6.1):

> key-p-tag = %x70 [FWS] "=" [ [FWS] base64string ]
>
>INFORMATIVE NOTE: A base64string is permitted to include white
>space (FWS) at arbitrary places; however, any CRLFs must be
>followed by at least one WSP character. Implementors and
>administrators are cautioned to ensure that selector TXT
>records conform to this specification.

**Comment by RvdHout** [comment]
Are you sure anything you have written is true?
I seriously doubt it, never experienced anything you describe above (even with 2048 bits keys, which are normally split up in multiple string by DNS

Few points of interest:
I notice you don't have any spaces between tags in your DKIM TXT record
I also notice you don't have a k=rsa tag in your DKIM TXT tag record, not required but very common to include

**Comment by pik-256** [comment]
I am sure but I feel I need to explain more: wyborcza.pl is a big polish newspaper publisher, it is not my domain. I cannot change anything on their domain.
I noticed a DKIM violation in all messages coming from this sender to my server. I mean: "X-hMailServer-Reason-1: Rejected by DKIM. - (Score: 3)" header in all incoming messages.
But a DKIM verifier plugin I have installed to Thunderbird claimed DKIM to be valid. I verified it and found the hmail debug message "DKIM: Unable to base64 decode public key found in DNS record".
However, I have posted it 6 months ago. In the meantime wyborcza.pl changed its key and now there are no spaces in it.
I tried to catch any current message with similar problems but I waited a week and got nothing, so I think very rarely anyone puts spaces in a base64 encoded DNS keys. Every day I receive several thousand messages on my hmail server. I will wait next week and if I do not catch pattern like this I'll try to set up a test environment to show the error.
